### ISSUE TITLE
Why process format if type validation fails

### ISSUE THREAD

**Issue by whitlockjc** [zaggino/z-schema#125]
Imagine I've got a schema like this:

```js
{
  "type": "integer",
  "format": "int32"
}
```

I register my `int32` format validator and I validate the value `1.1`.  It comes to my surprise that instead of getting a `type` error alone, I get a `type` and a `format` error:

```js
[
  {
    code: 'INVALID_TYPE',
    message: 'Expected type integer but found type number',
    path: []
  },
  {
    code: 'INVALID_FORMAT',
    message: 'Object didn\'t pass validation for format int64: 1.1',
    path: []
  }
]
```

That being said, why are we doing `format` validation on values that have failed `type` validation?

**Comment by zaggino** [comment]
yes, this should be fixed

**Comment by whitlockjc** [comment]
I searched but didn't see a reported issue, my bad.

**Comment by zaggino** [comment]
Hi @whitlockjc , I put this test in place: https://github.com/zaggino/z-schema/blob/master/test/ZSchemaTestSuite/Issue125.js
But it seems it's working fine, without chaning the code of the validator. Any change you're using the `breakOnFirstError: false` setting?

**Comment by whitlockjc** [comment]
I'm not using `breakOnFirstError` so whatever its default is.  Maybe I've got an old version of z-schema.  Here is my test: https://github.com/apigee-127/sway/blob/master/test/test-2.0.js#L272  Looks like I'm using `3.12.0` of z-schema.  Maybe it was just fixed after that.

**Comment by zaggino** [comment]
`3.12.2` is the latest now, but I don't think I have changed anything to fix this.

Anyway, please make a PR to this test: https://github.com/zaggino/z-schema/blob/master/test/ZSchemaTestSuite/Issue125.js if you think I got it wrong somewhere. Closing this in the meantime.

**Comment by whitlockjc** [comment]
I cannot reproduce this with standalone z-schema.  :)  Seems my project is somehow busting things up.  Not sure how but it is.

**Comment by whitlockjc** [comment]
This is really baffling.  Standalone z-schema works fine, z-schema in my environment fails.  If you get a few minutes: https://github.com/apigee-127/sway/blob/master/test/test-2.0.js#L303  That test, and most of the ones in that same `describe` block, all have the same type+format error issue.  If you've got a few minutes:

* `git clone https://github.com/apigee-127/sway.git`
* `cd sway`
* `npm install`
* `gulp test-node`

That shouldn't take more than a few minutes.  If you see these tests pass, we have some sort of situation in my usage of z-schema where it violates the contract your new test show works in standalone z-schema.

**Comment by zaggino** [comment]
```
  174 passing (4s)

[10:05:47] Finished 'test-node' after 4.99 s
```

**Comment by zaggino** [comment]
Can you make a test fail so I can check what's wrong with z-schema in your case?

**Comment by whitlockjc** [comment]
Here is a diff you can apply:

```diff
diff --git a/test/test-2.0.js b/test/test-2.0.js
index 71638c8..a5c103e 100644
--- a/test/test-2.0.js
+++ b/test/test-2.0.js
@@ -314,11 +314,6 @@ describe('sway (Swagger 2.0)', function () {
               code: 'INVALID_TYPE',
               message: 'Expected type integer but found type number',
               path: []
-            },
-            {
-              code: 'INVALID_FORMAT',
-              message: 'Object didn\'t pass validation for format int32: 1.1',
-              path: []
             }
           ]);
         });
@@ -373,11 +368,6 @@ describe('sway (Swagger 2.0)', function () {
               code: 'INVALID_TYPE',
               message: 'Expected type integer but found type number',
               path: []
-            },
-            {
-              code: 'INVALID_FORMAT',
-              message: 'Object didn\'t pass validation for format int64: 1.1',
-              path: []
             }
           ]);
         });

```

**Comment by zaggino** [comment]
Ah, error was completely in a different place, fixed in `3.12.3`

**Comment by whitlockjc** [comment]
Glad I wasn't making it up.  :)  Thanks for your help.

**Comment by frederikprijck** [comment]
Hey @zaggino,

Sorry to revive this, but I noticed the same behavior is introduced again, I am running v6.0.1:

```
const ZSchema = require("z-schema");

ZSchema.registerFormat("test", (str) => {
  return typeof str === 'string' &&  str.indexOf("https") > -1;
});

const validator = new ZSchema();

const schema = {
  type: "object",
  properties: {
    callbacks: {
      type: "array",
      items: {
        type: "string",
        format: "test",
      },
    },
  },
};

validator.validate(
  {
    callbacks: [true],
  },
  schema,
  (data, err) => {
    console.log(data);
  }
);
```
This outputs:
```
[
  {
    code: 'INVALID_TYPE',
    params: [ 'string', 'boolean' ],
    message: 'Expected type string but found type boolean',
    path: '#/callbacks/0',
    schemaId: undefined,
    [Symbol(z-schema/schema)]: { format: 'test', type: 'string', '__$validated': true },
    [Symbol(z-schema/json)]: { callbacks: [Array] }
  },
  {
    code: 'INVALID_FORMAT',
    params: [ 'test', true ],
    message: "Object didn't pass validation for format test: true",
    path: '#/callbacks/0',
    schemaId: undefined,
    [Symbol(z-schema/schema)]: { format: 'test', type: 'string', '__$validated': true },
    [Symbol(z-schema/json)]: { callbacks: [Array] }
  }
]
```
Is this expected in v6?

**Comment by zaggino** [comment]
@frederikprijck Published a fix in z-schema@6.0.2

**Comment by frederikprijck** [comment]
Confirmed, that works 👍 Thank you 🙏
